### PR TITLE
[BOLT] Fix C++ exceptions when LPStart is specified

### PR DESCRIPTION
**Pull request by maksfb**
Whenever LPStartEncoding was different from DW_EH_PE_omit, we used to miscalculate LPStart. As a result, landing pads were assigned wrong addresses. Fix that.

### CODE REVIEW

**Comment by aaupov** [comment]
The new test failed in pre-merge checks.

**Comment by maksfb** [comment]
Yes. But passing locally…

File: bolt/test/runtime/X86/exceptions-lpstart-zero.s [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+# RUN: %t.exe.bolt
+
+# REQUIRES: system-linux
+
+## Test that BOLT properly handles LPStart when LPStartEncoding is different
+## from DW_EH_PE_omit.
+
+# Test case compiled from:
+#
+# int main() {
```
Comment (ayermolo) [inline bolt/test/runtime/X86/exceptions-lpstart-zero.s:12]: what was the build command to produce main code and exception table?

File: bolt/lib/Core/Exceptions.cpp [hunk 2]
```diff
@@ -175,38 +180,38 @@ void BinaryFunction::parseLSDA(ArrayRef<uint8_t> LSDASectionData,
...
       outs() << "  current offset is " << (CallSitePtr - CallSiteTableStart)
              << '\n';
     }
 
     // Create a handler entry if necessary.
     MCSymbol *LPSymbol = nullptr;
-    if (LPOffset) {
+    if (LandingPad) {
+      // Verify if landing pad code is located outside current function
+      // Support landing pad to builtin_unreachable
```
Comment (ayermolo) [inline bolt/lib/Core/Exceptions.cpp:199]: Not sure I follow.

Comment (maksfb) [reply bolt/lib/Core/Exceptions.cpp:199]: If you mean the comment, it's moved from a different code location.

File: bolt/test/runtime/X86/exceptions-lpstart-zero.s (hunk 1 above)
Comment (maksfb) [reply bolt/test/runtime/X86/exceptions-lpstart-zero.s:12]: `clang input -o output.s -S`

**Comment by maksfb** [comment]
FYI: the reason the test is failing is unrelated to exceptions or the changes. Will post a separate PR.

**Comment by maksfb** [comment]
Rebuilt the test case with `-O1` to get rid of the `__builtin_unreachable()` problem described in #72899.

**Review by aaupov — APPROVED** [review]

**Pull request merged**